proxy_ring_buffer.go

```diff
@@ -0,0 +1,266 @@
+// uses the Ringbuffer for the first time. The most reliable way to do that is to write a dummy item into the Ringbuffer and
+// use the returned sequence as initial  id. On the reading side, this dummy item should be discarded. Please keep in mind that
+// this id is not the sequence of the item you are about to publish but from a previously published item. So it can't be used
+// to find that item.
+//
+// Add returns the sequence number of the added item. You can read the added item using this number.
+func (rb *Ringbuffer) Add(ctx context.Context, item interface{}, overflowPolicy OverflowPolicy) (sequence int64, err error) {
+	elementData, err := rb.validateAndSerialize(item)
+	if err != nil {
+		return ReadResultSetSequenceUnavailable, err
+	}
+	request := codec.EncodeRingbufferAddRequest(rb.name, int32(overflowPolicy), elementData)
+	response, err := rb.invokeOnPartition(ctx, request, rb.partitionID)
+	if err != nil {
+		return ReadResultSetSequenceUnavailable, err
+	}
+	return codec.DecodeRingbufferAddResponse(response), nil
+}
+
+// AddAll
+// Adds all the items of a collection to the tail of the Ringbuffer. A addAll is likely to outperform multiple calls
+// to add(Object) due to better io utilization and a reduced number of executed operations. If the batch is empty,
+// the call is ignored. When the collection is not empty, the content is copied into a different data-structure.
+// This means that: after this call completes, the collection can be re-used. the collection doesn't need to be serializable.
+// If the collection is larger than the capacity of the Ringbuffer, then the items that were written first will be
+// overwritten. Therefor this call will not block. The items are inserted in the order of the Iterator of the collection.
+// If an addAll is executed concurrently with an add or addAll, no guarantee is given that items are contiguous.
+// The result of the future contains the sequenceId of the last written item.
+func (rb *Ringbuffer) AddAll(ctx context.Context, overflowPolicy OverflowPolicy, items ...interface{}) (int64, error) {
+	elementData, err := rb.validateAndSerializeValues(items)
+	if err != nil {
+		return ReadResultSetSequenceUnavailable, err
+	}
+	request := codec.EncodeRingbufferAddAllRequest(rb.name, elementData, int32(overflowPolicy))
+	response, err := rb.invokeOnPartition(ctx, request, rb.partitionID)
+	if err != nil {
+		return ReadResultSetSequenceUnavailable, err
+	}
+	return codec.DecodeRingbufferAddAllResponse(response), nil
+}
+
+// ReadOne
+// Reads one item from the Ringbuffer. If the sequence is one beyond the current tail, this call blocks until an
+// item is added. This method is not destructive unlike e.g. a queue.take. So the same item can be read by multiple
+// readers, or it can be read multiple times by the same reader. Currently, it isn't possible to control how long this
```

Comment:
I would love to discuss this point. The context parameter provides a way to pass a value on how much it will block. Context cancelation and timeout will result in an error, but it is doable for the user.

Do you guys agree? If yes, how can we document this 🤔

Reply:
Handling timeouts explicitly I consider best practice ... I do suggest adding an example.

Reply:
I did add some example code
2bdad54129f947b5b8a7b4d2c80d7a185ac69878
What do you think?

Reply:
Thanks for the example! I think it looks good but I'm not sure if we should support it with context since other clients do not have such a mechanism yet. I would love to hear other reviewers' thoughts as well on this. If that's okay for you, can you leave this conversation as "unresolved" for now?

Reply:
> The context parameter provides a way to pass a value on how much it will block. Context cancelation and timeout will result in an error, but it is doable for the user.

What do you mean by an error? So it is possible to set timeouts and cancel the API call in go by default ? Well, I did not understand the problem. I don't think there will be any harm of supporting them with context because we have context (as far as I know) in all async APIs in the go client

Reply:
Go client supports context cancellation and timeouts natively, which is different from other clients. I think we can just remove the "Currently, it isn't possible to control how long this  call is going to block. In the future we could add e.g. tryReadOne(long sequence, long timeout, TimeUnit unit)." part from the documentation.

Reply:
I could not found the statement of "Go client supports context cancellation and timeouts natively" in `pkg doc` or in official repository. We are only saying that;
`Go context support for all distributed data structures.`. User can assumes that we are also supporting `context.WithDeadline` based on this statement.

I think context related edge cases and usage should be explained in more verbose way, especially for the public API.

Reply:
> Go client supports context cancellation and timeouts natively, which is different from other clients. I think we can just remove the "Currently, it isn't possible to control how long this call is going to block. In the future we could add e.g. tryReadOne(long sequence, long timeout, TimeUnit unit)." part from the documentation.

Since there is no additional error type other than the `Cancelled` and `DeadlineExceeded`, can't we say that the public API calls support context usage in any sense without denoting the cancellation and timeout ? Because, user can use `WithDeadline` and it is quite similar to `WithTimeout` and produces same `DeadlineExceeded` error.

Reply:
@nevzatseferoglu @yuce @srknzl 
Since Go supports timeouts out of the box, may I propose, that we simply remove this from the documentation?
I took the freedom to commit it already, please do veto if you intent otherwise.
See bdf427baf5a29bae7beb9e3c76136f9dcf857f1d